source/guide/3.2/guide-security.md

```diff
@@ -11,33 +11,230 @@ pageord: 500
+Of course, direct requests made using other clients (e.g. curl) must also include correct credentials, or they will
+fail with an "Unauthorized User" error.
+
 
 # Main Concepts
 
-## Userstores
+## Userstore
+Generally, a userstore is simply a class that enables loading of user details and returns them as a user object.
+
+Typically (but not always) user details are stored as records in a database or objects in a directory. Each user can
+be identified by a unique attribute, such as a username or id, or by a unique combination of attributes.<br/>
+In order to authenticate a user (for example by a set of username and password) it might be required to load the user's
+details and verify the given credentials indeed match.<br/>
+To support a variety of user-store systems and configurations Cloudify security framework can accept different
+userstore implementations. It's possible to use the default Flask-secuREST simple userstore or to specify a new
+implementation that supports a specific userstore system.
+
+A valid userstore implementation can be any Python class that inherits from
+[abstract_userstore.py](https://github.com/cloudify-cosmo/flask-securest/blob/master/flask_securest/userstores/
+abstract_userstore.py) and implements a `get_user` method.
+The object returned by `get_user` must adhere to Flask-secuREST's User Model, specifying these 3 methods:
+
+- is_active()
+- is_anonymous()
+- get_roles()
+
+
+## Authentication Provider
+An Authentication Provider is a class that performs authentication. Multiple authentication providers can be configured
+in order to support different authentication methods (e.g. password, token, Kerberos).<br/>
+When a REST call is received by the REST service, the security framework will attempt to authenticate it using each
+authentication provider in the order of configuration, until authentication succeeds.
+The authentication provider has access to the userstore instance (if configured) and can use it to get user details and
+use them to perform authentication.<br/>
+For example, it can compare the given password to the one found on the userstore or verify the user is still active
+(in many environments users are marked as "inactive", instead of deleting the account entirely).
+
+
+Once an authenticator can successfully authenticate the request's user - it should return the user object and allow the
+request to be completed. Other authenticators will not be called until the next request is processed.
+If non of the authenticators authenticates the request - the request will not reach its endpoint and the client will
+receive an "Unauthorized User" error.
+
+A valid authentication provider implementation can be any Python class that inherits from
+[abstract_authentication_provider.py](https://github.com/cloudify-cosmo/flask-securest/blob/master/flask_securest/
+authentication_providers/abstract_authentication_provider.py) and implements an `authenticate` method.
+
```

Comment:
kinda unclear about the signature of the method itself - from what i understand, if i want to return a success, then the method should return the user object. what i should i do otherwise? raise an error? return false or none?